broadinstitute/single_cell_portal_core
# Prototype: Validate files in client: reduce time to fail from 2-3 min to < 1 s (SCP-3496, SCP-3498)

This adds a prototype of basic client-side validation for study files.  It has two big benefits:

* Reduce time to fail from 2-3 minutes to less than 1 second
* Report errors within app UI instead of via email

For now, this feature is hidden by default behind the flag `clientside_validation`.  Work remains for additional validation rules, error reporting UI, and analytics.

The implementation uses the `FileReader` API to read a file that the user selects from their local file system.  The file is then validated using JavaScript that closely follows patterns from the Python implementation of validation rules in Ingest Pipeline (entry point e.g. [here](https://github.com/broadinstitute/scp-ingest-pipeline/blob/development/ingest/annotations.py#L328)).

To test:
1. Do migration: run `ruby rails_local_setup.rb`, then `source config/secrets/.source_env.bash`, then `rails db:migrate`
2. Sign in
3. Open upper right nav menu -> Admin Controls -> User Roles (https://localhost:3000/single_cell/admin#users)
4. Find yourself, set `clientside_validation` feature flag to "True"
5. Create a new study, e.g. "Client-side validation test"
6. In "Metadata" tab of upload wizard, click "Choose File"
7. Select file at `single_cell_portal_core/test/test_data/metadata_example_bad_TYPE.txt`
8. Verify that you see an alert reporting an error, as shown in brief video below.


https://user-images.githubusercontent.com/1334561/128394912-1213fa03-042c-4d43-bafb-a128da94d506.mov



This satisfies SCP-3496 and SCP-3498.
